### ISSUE TITLE
AR Mass Edit

### ISSUE THREAD

**Issue by ramonski** [senaite/senaite.core#300]
## Steps to reproduce

Create multiple ARs and try to edit them.

## Current behavior

The user has to navigate to each AR to edit the base data or manage Analyses.

## Expected behavior

Multiple ARs can be edited with a similar form like used in the new "add form".

## Screenshot (optional)

<img width="952" alt="multi-edit" src="https://user-images.githubusercontent.com/713193/31980656-46b22604-b94e-11e7-95d9-dfae447879ff.png">

**Comment by lemoene** [comment]
All of this can be easily achieved, an in very rich functionality, trimmed down for merging with Bika LIMS 3.4, by merging @rockfruit's splendid UW code, and also do away with Bulk AR Import (by using the UW Batch AR code)

Thanks for putting me onto this one @grulisco

**Comment by ramonski** [comment]
@rockfruit: Can you please technically explain what you have done in the UW branch?

The the following requirements are behind this feature:

- Based on AR Add 2 to avoid code duplication
- Capability to add/remove Profiles
- Allow fields to be edited depending on the WF state of the AR

Until now this was only a PoC (see PR), so it would be great to have this ready to use in code.
A related issue in the old tracker was this one: https://github.com/bikalims/bika.lims/issues/2192

**Comment by lemoene** [comment]
Is their progress on this. If we can form a reasonable time and effort estimate, i'll pump for sponsorship of the Create part to start off with. Please post your estimates (and pointers to the code) @rockfruit, @ramonski.

We are migrating AR Import for Cannabis and once we have that in place as a fall back, could turn to Batch AR create if it is not too big a job. Shouldn't be, it worked for UW

**Comment by ramonski** [comment]
I'm still waiting on an answer on this @lemoene. If I understood you right, this feature is already done in the UW branch? If so, please add a PR for review. Thanks

**Comment by lemoene** [comment]
K. Time and effort @mikejmets ?

**Comment by lemoene** [comment]
I did not dig far enough for the UW comms to retrieve a spec, but _ It was functional and trimmed down_ from elaborate UW spec _for merging with Bika LIMS 3.4 already_ if I iunderstood @rockfruit correctly. Before that, a review ran beautifully

see if I can find a spec...

**Comment by lemoene** [comment]
Found an old running instance, not the latest, where this and also Batch and WS integrations, reach useful status, https://jira.bikalabs.com/browse/LIMS-2750

It completes quite a bit already, though clumsily, I thought we had better UI, maybe in the final versions

![image](https://user-images.githubusercontent.com/1196034/35858974-010c28de-0b47-11e8-87fa-1bcca7685224.png)

**Comment by ramonski** [comment]
@lemoene: You were talking that you have ["beautiful code for this, and it is beyond you why it gets ignored"](https://github.com/orgs/senaite/teams/community/discussions/5?from_comment=2#discussion-5-comment-2). The link you posted above shows just numerous screenshots, but I can find neither a specification nor the ["splendid UW code"](https://github.com/orgs/senaite/teams/community/discussions/5?from_comment=2#discussion-5-comment-2) you mentioned.

The screenshot above shows the AR Add form as the label already suggests. So I don't understand that neither.

Please provide that code either as a link or in a PR (preferred), and please only use the tracker for technical related contributions and use other communication channels for time/effort/sponsorship discussions – Thanks for understanding.

**Comment by lemoene** [comment]
As you wish, you leave the project vulnerable to duplication by not allowing discussions like this. It happened already. There are not nearly enough LIMS experience in Riding Bytes to discourage rich user and business analyst discussions in a tracker of some sorts. It already cost us dearly when you announced Senaite production-ready when it was only for a singular use case.

IMO you should be using all the user and analyst feedback on offer and exploit it. Maybe you saw the Beekeeper model. You should also state clearly that this forum is for coders only, I won't do it until a suitable replacement is set up. Frustrating that you simply cannot let it be and fro self regulation to rule.

I have been talking about this really valuable cache of UW code for months, am not tech and referred you to  proud author Campbell and stepped on it again when you met in Barcelona. It was widely ignored. This is not the platform for that discussion, but he got badly undermined in his attitude towards me, what little grip I had on his wayward genius evaporated and we are reaping the consequences now.

Wrt to this issue, I clearly laid out that I don't have my hands on the code and the screenshots were used by way of illustration of what was achieved already. The obvious place to do that was on this issue here before things gets duplicated. That is also the only spec I have it seems, apart from initial older ones - I came to the conclusion that it was finalised in many granular improvements requests in the Bika tracker. I can write it up, but then there is nowhere to stick it up.

Also not for here, but the Senaite space is clearly not shaped optimally or entirely free and already some of our work that should be promoting Senaite gets painted in at Bikalims.org. Decision time is fast approaching whether maybe to continue under that banner where we can operate in an uninhibited environment. Bika Cannabis, because of the hype around it, and the alliteration in the name!-) will be marketed as such.

**Comment by lemoene** [comment]
annotating the screenshot

**Comment by xispa** [comment]
UW code always lived in its own fork, as well as baobab, cannabis and others. Obviously, one cannot expect Senaite to include functionalities that were disseminated amongst a myriad of custom-made forks that were meant to be used for specific clients only. Claiming here that we missed (and/or we don't want) to merge that "splendid" code that no one knows where it currently lives is non-sense.

If you are able to find that code and think is worth, open a Pull Request, but please @lemoene , don't add more confusion with functionalities that never reached the public repos.

**Comment by lemoene** [comment]
To clear the confusion I'll continue with tracking feature requests in Jira, there is great value to the information collected there already, will clean it up. That could help preventing duplication somewhat.

The UW project was very public, I only took their tracker items private recently when they were considering a restart. Maybe the code ended up in Campbell's not easy to find. LIMS analysts and users hardly go there.

The functionality was Bika LIMS 3.4 ready, communicated again before the Barcelona meet-up. The migration ship was sailing just them. Indeed I came to now realise it is an opportunity missed, great  functionality, and stable. No feedback was given by anybody, 'not wanting to' I did not say, just thought 'too busy'. 

The functions are now probably better rewritten. I was holding out for maybe more affordable refactoring, having at a minimum a proto-type and some code to work from.

**Comment by rockfruit** [comment]
I must chime in here (better late than never).  I've just now written and deleted an essay on this discussion!; however this comment will instead focus on the subject of this issue, **AR Mass Edit**, which is functionality that has never existed in Bika LIMS before.

I agree that the quickest, simplest, and most robust solution is to use the add2 form.  It's perfect.

The only place this breaks down a bit is when a user wants to change one field value inside some hundreds of ARs simultaneously.  There was a function to switch layout of the AR-Add form from columns to rows; combined with the ability to modify which fields are displayed in the form, this may allow the add2 form to mass-edit very many ARs.

##

Side note regarding "arimport": if the add2 form supports switching layout, it may well be used as an intermediate stage for importing ARs from an externally created file; however I feel that's a cop-out.  If our UI is so bad that a user must resort to an XLS file, then our job is to create a better UI.  The row-layout add2 form could be sufficient.

**Comment by ramonski** [comment]
Won't be implemented.
